src/utils/Render.js

```diff
@@ -115,23 +144,204 @@ function renderWithStrategy(strategy) {
     // cache the element
     const Element = React.createElement(Container, props)
 
-    // render so we kick things off and get the props
-    buffer.fill(Element)
+    const start = Date.now()
 
-    // flush out the results in the buffer synchronously setting the store
-    // state and returning the markup
-    return buffer.flush(alt, Element)
+    return buffer.render(alt, Element, info).then((obj) => {
+      const time = Date.now() - start
+
+      return {
+        error: obj.error,
+        html: obj.html,
+        state: obj.state,
+        buffer: obj.buffer,
+        element: obj.element,
+        diagnostics: {
+          iterations: obj.diagnostics.iterations,
+          dispatches: obj.diagnostics.dispatches,
+          time
+        }
+      }
+    })
   }
 }
 
-export function toDOM(Component, props, documentNode, shouldLock) {
-  const buffer = new DispatchBuffer()
-  buffer.locked = !!shouldLock
-  const Node = usingDispatchBuffer(buffer, Component)
-  const Element = React.createElement(Node, props)
-  buffer.clear()
-  return React.render(Element, documentNode)
-}
+export default class Render {
+  constructor(alt, options = {}) {
+    this.alt = alt
+    this.options = options
+
+    // defaults
+    // 500ms or 5 iteration max
+    this.options.timeout = options.timeout || 500
+    this.options.maxIterations = options.maxIterations || 5
+  }
+
+  toString(Component, props) {
+    this.options.i = 0
+    this.options.time = 0
+    return renderWithStrategy('renderToString')(
+      this.alt,
+      Component,
+      props,
+      this.options
+    )
+  }
+
+  toStaticMarkup(Component, props) {
+    this.options.i = 0
+    this.options.time = 0
+    return renderWithStrategy('renderToStaticMarkup')(
+      this.alt,
+      Component,
+      props,
+      this.options
+    )
+  }
+
+  getReady(Component, props, opts = {}) {
+    const buffer = new DispatchBuffer()
+
+    if (opts.status) buffer.status = opts.status
+    const Node = usingDispatchBuffer(buffer, Component)
+    const Element = React.createElement(Node, props)
+    buffer.clear()
+    return Promise.resolve(Element)
+  }
+
+  toDOM(Component, props, documentNode, opts = {}) {
+    return this.getReady(Component, props, opts).then((Element) => {
+      return React.render(Element, documentNode)
+    })
+  }
+
+  static connect(Spec, MaybeComponent) {
```

Comment:
I don't know that there's a direct analogy here, but e.g. I've very intentionally made http://martyjs.org/api/inject-app/index.html usable as a decorator, while http://martyjs.org/api/containers/index.html is not - the former modifies the class in-place and won't break the use cases I referenced above, while the latter adds a container component, and is likely to do so.